lib/web_ui/lib/src/engine/canvaskit/text_fragmenter.dart

```diff
@@ -0,0 +1,91 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+
+import '../dom.dart';
+import '../text/line_breaker.dart';
+
+abstract class CkTextFragmenter {
+  const CkTextFragmenter(this.text);
+
+  /// The text to be fragmented.
+  final String text;
+
+  /// Performs the fragmenting of [text] and returns a [Uint32List].
+  Uint32List fragment();
+}
+
+/// The granularity at which to segment text.
+///
+/// To find all supported granularities, see:
+/// - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter/Segmenter
+enum IntlSegmenterGranularity {
+  grapheme,
+  word,
+}
+
+class CkIntlFragmenter extends CkTextFragmenter {
+  CkIntlFragmenter(super.text, this.granularity)
+      : assert(domIntl.Segmenter != null);
+
+  final IntlSegmenterGranularity granularity;
+
+  String get _granularityString {
+    switch (granularity) {
+      case IntlSegmenterGranularity.grapheme:
+        return 'grapheme';
+      case IntlSegmenterGranularity.word:
+        return 'word';
+    }
+  }
+
+  @override
+  Uint32List fragment() {
+    final DomIteratorWrapper<DomSegment> iterator =
+        createIntlSegmenter(granularity: _granularityString)
```

Comment:
Same question here. Is `IntlSegmenter` stateful? Can we memoize one for each granularity and avoid creating them on every string? This will become more relevant with wasm, where we probably want to avoid too many wasm-js hops.

Reply:
Done.

Regarding the wasm thing: I don't think this memoization makes a difference. The majority of chatting with js happens *after* creating the `Intl.Segmenter` (calling `segment()` and iterating over the result). If it becomes an issue, we can implement a JS function that wraps this entire block of code and does all the chatting on the JS side and returns the final result (i.e. `Uint32List`) to wasm. Not sure how this can be done in practice with dart2wasm.

cc @eyebrowsoffire to keep in mind that this might be needed for the skwasm project. It would be nice to have some kind of annotation on methods to tell the compiler that it's better to keep the entire method in JS and not compile it to wasm. I may write some thoughts around this in a new issue.